Rewrite ping

arcusfelis opened the pull request:
Changes:
- Introduce cets_ping with better node pinging strategy. It interrupts ping, if pang is received. It does not use net_adm:ping/1 to avoid going though net_kernel and potentially blocking it for a couple of seconds if the remote node is not available (couple of seconds adds up fast, if we have a lot of nodes). The current strategy is try to call EPMD first to check if the remote address is resolvable.
- Improve logging. Report nodeup/nodedown with downtime duration and time since the node start. It is especially useful when tweaking probes and upgrading k8s.
- Added node_down_history field to CETS. It looks like currently unused, but the purpose is to be able to analyse it manually and check after netsplits. It is added, because before we had conflict_nodes/conflict_tables persistently (i.e. after a netsplit, nodes see different lists). Still, I cannot reproduce conflict bug anymore, after tens of runs - so works as expected.

arcusfelis committed:
Improve net_adm:ping/1

Do not call net_kernel to connect to a node, if its name is not resolvable
Do not disconnect from a node if ping fails

arcusfelis committed:
Rewrite ping logic - use connect_node there instead

Add ping_pairs function
ping_pairs does not use net_adm:ping anymore

arcusfelis committed:
Do check_could_reach_each_other before check_fully_connected

Do pings before sending any messages to new nodes

arcusfelis committed:
Report transaction retry on info level, not error

arcusfelis committed:
Log nodeup/nodedown events with warning level

arcusfelis committed:
Report downtime duration

arcusfelis committed:
Use undefined instead of unknown for timestaps in logs

arcusfelis committed:
Skip undefined timestamps in logs

arcusfelis committed:
Report alive nodes count on nodeup/nodedown

arcusfelis committed:
Report time_since_startup_in_milliseconds for nodeup/nodedown events

arcusfelis committed:
Use epmd instead of resolver to test if node is pingable

arcusfelis committed:
Do not report long task failing with reason stop

arcusfelis committed:
Log warning if same node reconnects

arcusfelis committed:
Add node_down_history

arcusfelis committed:
Sort available nodes in the status API

arcusfelis committed:
Add +node_down_history_is_updated_when_netsplit_happens testcase

arcusfelis committed:
Add disco_logs_nodeup

arcusfelis committed:
Add disco_logs_nodedown

arcusfelis committed:
Add disco_logs_nodeup_after_downtime testcase

arcusfelis committed:
Move logging into handle_nodedown/handle_nodeup

arcusfelis committed:
Add disco_logs_node_reconnects_after_downtime testcase

arcusfelis committed:
Add disco_logs_nodeup_no_log/disco_logs_nodedown_no_log testcases

arcusfelis committed:
Tweak for code coverage and logging

arcusfelis committed:
Test that start_time is set in logs

arcusfelis committed:
Add disco_nodeup_timestamp_is_updated_after_node_reconnects testcase

arcusfelis committed:
Add disco_node_start_timestamp_is_updated_after_node_restarts testcase

arcusfelis committed:
Add cets_discovery into a list of modules to disable logging during testing

There is a special test group for that.
So, we test with logging enabled and disabled.

arcusfelis committed:
Use proto_dist to just do one adress_please call per ping

arcusfelis committed:
Add cets_ping_non_existing_node testcase

arcusfelis committed:
Add tests for net_family code in cets_ping

arcusfelis committed:
Add stop_reason_is_not_logged_in_tracked testcase

arcusfelis committed:
Fail if cannot split nodename in cets_ping

arcusfelis committed:
Fix flaking tests

arcusfelis committed:
Rename special reason from stop to shutdown

arcusfelis committed:
Add unexpected_nodedown_is_ignored_by_disco testcase

arcusfelis committed:
Add ping_pairs tests

chrzaszcz commented on src/cets_ping.erl:
```diff
@@ -0,0 +1,86 @@
+            inet
+    end;
+net_family(_) ->
+    inet.
```
This can be simplified, because the ones for IPv6 start with `inet6`:
```suggestion
net_family({ok, [["inet6" ++ _]]}) ->
    inet6;
net_family(_) ->
    inet.
```

chrzaszcz commented on src/cets_ping.erl:
```diff
@@ -0,0 +1,86 @@
+    %%   (because it could disconnect already connected node because of race conditions)
+    %% - it calls net_kernel's gen_server of the remote server,
+    %%   but it could be busy doing something,
+    %%   which means slower response time.
```
```suggestion
    %% We could use net_adm:ping/1 but it:
    %% - disconnects node on pang - we don't want that
    %%   (because it could disconnect an already connected node because of race conditions)
    %% - calls net_kernel's gen_server of the remote server,
    %%   but it could be busy doing something, which means slower response time.
```

chrzaszcz commented on src/cets_ping.erl:
```diff
@@ -0,0 +1,86 @@
+        Other ->
+            %% We do not need to ping the rest of nodes -
+            %% one node returning pang is enough to cancel join.
+            %% We could exit earlier and safe some time
```
```suggestion
            %% We could exit earlier and save some time
```

chrzaszcz commented on test/cets_SUITE.erl:
```diff
@@ -1009,6 +1029,45 @@ atom_error_is_logged_in_tracked(_Config) ->
+    wait_for_down(Pid),
+    [] = receive_all_logs_with_log_ref(?FUNCTION_NAME, LogRef).
+
+%% Counter example for shutdown_reason_is_not_logged_in_tracked
```
I'd rather change it to something like this:
```suggestion
%% Complementary to shutdown_reason_is_not_logged_in_tracked
```

chrzaszcz commented on test/cets_SUITE.erl:
```diff
@@ -2208,10 +2264,340 @@ join_interrupted_when_ping_crashes(Config) ->
+    Disco = start_disco(Node1, #{
+        name => DiscoName, backend_module => cets_discovery_fun, get_nodes_fn => F
+    }),
+    cets_discovery:add_table(Disco, Tab),
```
Could you reuse this part, because it is the same as in the previous test case?

chrzaszcz commented on test/cets_SUITE.erl:
```diff
@@ -2208,10 +2264,340 @@ join_interrupted_when_ping_crashes(Config) ->
+    end.
+
+disco_node_down_timestamp_is_remembered(Config) ->
+    logger_debug_h:start(#{id => ?FUNCTION_NAME}),
```
This line is probably unecessary

chrzaszcz commented on test/cets_SUITE.erl:
```diff
@@ -2208,10 +2264,340 @@ join_interrupted_when_ping_crashes(Config) ->
+    }),
+    cets_discovery:add_table(Disco, Tab),
+    ok = cets_discovery:wait_for_ready(Disco, 5000),
+    rpc(Peer2, erlang, disconnect_node, [Node1]),
```
This part also repeats - I had to spend some time on comparing the two parts, and with a helper function it would be obvious.

chrzaszcz commented on test/cets_SUITE.erl:
```diff
@@ -2208,10 +2264,340 @@ join_interrupted_when_ping_crashes(Config) ->
+    }),
+    cets_discovery:add_table(Disco, Tab),
+    ok = cets_discovery:wait_for_ready(Disco, 5000),
+    rpc(Peer2, erlang, disconnect_node, [Node1]),
```
Why disconnect if we are checking `nodeup`, not `nodedown`? Is this line unnecessary? It looks like it was copied from the testcase below.

arcusfelis replied:
We check reconnect here.
After disconnect_node on this line, disco should reconnect automatically.
We selectively receive nodeup with downtime_millisecond_duration field.
downtime_millisecond_duration field would only get added after the second nodeup (first one we never extract from the message box).

If we remove this disconnect_node line, the test would fail with the timeout in the receive statement.

I add more comments here into the code though, describing what we test.

chrzaszcz commented on test/cets_SUITE.erl:
```diff
@@ -2208,10 +2264,340 @@ join_interrupted_when_ping_crashes(Config) ->
+    ).
+
+disco_node_start_timestamp_is_updated_after_node_restarts(Config) ->
+    logger_debug_h:start(#{id => ?FUNCTION_NAME}),
+    Node1 = node(),
+    #{ct2 := Peer2} = proplists:get_value(peers, Config),
+    #{ct2 := Node2} = proplists:get_value(nodes, Config),
+    Tab = make_name(Config),
+    {ok, _Pid1} = start(Node1, Tab),
+    {ok, _Pid2} = start(Peer2, Tab),
+    F = fun(State) ->
+        {{ok, [Node1, Node2]}, State}
+    end,
+    DiscoName = disco_name(Config),
+    Disco = start_disco(Node1, #{
+        name => DiscoName, backend_module => cets_discovery_fun, get_nodes_fn => F
+    }),
+    %% We have to start discovery server on both nodes for that feature to work
+    Disco2 = start_disco(Node2, #{
+        name => DiscoName, backend_module => cets_discovery_fun, get_nodes_fn => F
+    }),
+    cets_discovery:add_table(Disco, Tab),
+    ok = cets_discovery:wait_for_ready(Disco, 5000),
```
Unless I missed something (it's difficult to compare manually), this chunk repeats for 3 test cases. Could you extract it to a helper?

chrzaszcz reviewed (commented):
Looks good in general. The test suite is becoming lengthy, and I was getting lost in the test cases...

arcusfelis committed:
Simplify ipv6 matching in net_family

arcusfelis committed:
Fix typos in cets_ping

arcusfelis committed:
Deduplicate code in tests

Add setup_two_nodes_and_discovery helper function

arcusfelis committed:
Trigger get_nodes on nodeup

Add disco_nodeup_triggers_check_and_get_nodes testcase
There is a very high chance that there could be something new in the DB
on nodeup - so trigger the DB read. Useful if we wanna know IPs after nodeup
and IPs are stored in DB

arcusfelis committed:
Remove outdated comment

chrzaszcz reviewed (approved):
👌
